lazy-pages/src/sys.rs

```diff
@@ -39,117 +53,284 @@ cfg_if! {
+    log::trace!("mprotect r/w, addr = {unprot_addr:#x}, size = {unprot_size:#x}");
+    region::protect(
+        unprot_addr as *mut (),
+        unprot_size as usize,
+        Protection::READ_WRITE,
+    )?;
 
-    let unprot_size = gear_pages_num * gear_ps;
-
-    region::protect(unprot_addr as *mut (), unprot_size, Protection::READ_WRITE)?;
-
-    LAZY_PAGES_CONTEXT.with(|ctx| {
-        let mut ctx = ctx.borrow_mut();
-        for idx in 0..gear_pages_num as u32 {
-            let page = gear_page + idx.into();
-
-            let ptr = (unprot_addr as *mut u8).add(idx as usize * gear_ps);
-            let buffer_as_slice = std::slice::from_raw_parts_mut(ptr, gear_ps);
-
-            // TODO: simplify before release (issue #1147). Currently we must support here all old runtimes.
-            // For new runtimes we have to calc page key from program pages prefix.
-            let page_key = if let Some(prefix) = &ctx.program_storage_prefix {
-                page_key_in_storage(prefix, page)
-            } else {
-                // This case is for old runtimes support
-                ctx.lazy_pages_info.remove(&page).ok_or(Error::LazyPageNotExistForSignalAddr(mem, page))?
-            };
-            let res = sp_io::storage::read(&page_key, buffer_as_slice, 0);
-
-            if res.is_none() {
-                log::trace!(
-                    "{:?} has no data in storage, so just save current page data to released pages",
-                    page
-                );
-            } else {
-                log::trace!("{:?} has data in storage, so set this data for page and save it in released pages", page);
+    let fist_gear_page = wasm_addr.as_page_number();
+
+    for idx in 0..unprot_size / lazy_page_size {
+        // Arith opertaions are safe here.
+        let lazy_page_wasm_addr = wasm_addr.get() + idx * lazy_page_size;
+        let begin = fist_gear_page.0 + idx * num_of_gear_pages_in_one_lazy;
+        let end = begin + num_of_gear_pages_in_one_lazy;
+
+        if ctx.accessed_pages_addrs.contains(&lazy_page_wasm_addr) {
+            log::trace!("lazy page {lazy_page_wasm_addr:#x} is already accessed");
+            for gear_page in (begin..end).map(PageNumber) {
+                log::trace!("add {gear_page:?} to released");
+                if !ctx.released_pages.insert(gear_page) {
+                    return Err(Error::DoubleRelease(gear_page));
+                }
             }
+            continue;
+        }
+
+        ctx.accessed_pages_addrs.insert(lazy_page_wasm_addr);
+
+        for gear_page in (begin..end).map(PageNumber) {
+            let page_buffer_ptr = (wasm_mem_addr as *mut u8).add(gear_page.offset());
+            let buffer_as_slice = std::slice::from_raw_parts_mut(page_buffer_ptr, gear_ps as usize);
+            let res = sp_io::storage::read(prefix.calc_key_for_page(gear_page), buffer_as_slice, 0);
+
+            log::trace!(
+                "{:?} has{} data in storage",
```

Comment:
```suggestion
                "{:?} has data in storage: {}",
```
and then only `res.is_none()`

Just simpler

Reply:
Done